<div align="center">

  <a href="https://ultralytics.com/yolov5" target="_blank">
    <img width="1024", src="https://raw.githubusercontent.com/ultralytics/assets/master/yolov5/v70/splash.png"></a>


<br>
  <a href="https://bit.ly/yolov5-paperspace-notebook"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a>
  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/segment/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a>
  <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
<br>

This <a href="https://github.com/ultralytics/yolov5">YOLOv5</a> 🚀 notebook by <a href="https://ultralytics.com">Ultralytics</a> presents simple train, validate and predict examples to help start your AI adventure.<br>See <a href="https://github.com/ultralytics/yolov5/issues/new/choose">GitHub</a> for community support or <a href="https://ultralytics.com/contact">contact us</a> for professional support.

</div>

# Setup

Clone GitHub [repository](https://github.com/ultralytics/yolov5), install [dependencies](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) and check PyTorch and GPU.

In [1]:
!git clone https://github.com/ultralytics/yolov5  # clone
%cd yolov5
%pip install -qr requirements.txt  # install

import torch
import utils
display = utils.notebook_init()  # checks

YOLOv5 🚀 v7.0-33-g1752768 Python-3.8.10 torch-1.13.0+cu117 CUDA:0 (NVIDIA GeForce RTX 3080, 9996MiB)


Setup complete ✅ (32 CPUs, 31.3 GB RAM, 99.3/479.5 GB disk)


# 1. Predict

`segment/predict.py` runs YOLOv5 instance segmentation inference on a variety of sources, downloading models automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases), and saving results to `runs/predict`. Example inference sources are:

```shell
python segment/predict.py --source 0  # webcam
                             img.jpg  # image 
                             vid.mp4  # video
                             screen  # screenshot
                             path/  # directory
                             'path/*.jpg'  # glob
                             'https://youtu.be/Zgi9g1ksQHc'  # YouTube
                             'rtsp://example.com/media.mp4'  # RTSP, RTMP, HTTP stream
```

In [ ]:
!python detect.py --weights runs/train/exp10/weights/best.pt --img 640 --conf 0.1 --source ../yolo_images/test_data/test/Screenshot\ from\ 2022-12-13\ 19-04-45.png
# display.Image(filename='runs/predict/exp/Screenshot from 2022-12-13 18-57-29.png', width=600)

detect: weights=['runs/train/exp10/weights/best.pt'], source=../yolo_images/test_data/test/Screenshot from 2022-12-13 19-04-45.png, data=data/coco128.yaml, imgsz=[640, 640], conf_thres=0.1, iou_thres=0.45, max_det=1000, device=, view_img=False, save_txt=False, save_conf=False, save_crop=False, nosave=False, classes=None, agnostic_nms=False, augment=False, visualize=False, update=False, project=runs/detect, name=exp, exist_ok=False, line_thickness=3, hide_labels=False, hide_conf=False, half=False, dnn=False, vid_stride=1
YOLOv5 🚀 v7.0-33-g1752768 Python-3.8.10 torch-1.13.0+cu117 CUDA:0 (NVIDIA GeForce RTX 3080, 9996MiB)

Fusing layers... 
Model summary: 157 layers, 7023610 parameters, 0 gradients, 15.8 GFLOPs
image 1/1 /home/killuh/ws_p38/appium/src/notebooks/yolo_images/test_data/test/Screenshot from 2022-12-13 19-04-45.png: 640x320 1 login field, 2 password fields, 1 to login page, 1 next buttons, 6.6ms
Speed: 0.2ms pre-process, 6.6ms inference, 1.2ms NMS per image at shape (1, 3, 640

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<img align="left" src="https://user-images.githubusercontent.com/26833433/199030123-08c72f8d-6871-4116-8ed3-c373642cf28e.jpg" width="600">

# 2. Validate
Validate a model's accuracy on the [COCO](https://cocodataset.org/#home) dataset's `val` or `test` splits. Models are downloaded automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases). To show results by class use the `--verbose` flag.

In [ ]:
# Download COCO val
!bash data/scripts/get_coco.sh --val --segments  # download (780M - 5000 images)

######################################################################## 100.0%
######################################################################## 100.0%


In [ ]:
# Validate YOLOv5s-seg on COCO val
!python segment/val.py --weights yolov5s-seg.pt --data coco.yaml --img 640 --half

segment/val: data=/content/yolov5/data/coco.yaml, weights=['yolov5s-seg.pt'], batch_size=32, imgsz=640, conf_thres=0.001, iou_thres=0.6, max_det=300, task=val, device=, workers=8, single_cls=False, augment=False, verbose=False, save_txt=False, save_hybrid=False, save_conf=False, save_json=False, project=runs/val-seg, name=exp, exist_ok=False, half=True, dnn=False
YOLOv5 🚀 v7.0-2-gc9d47ae Python-3.7.15 torch-1.12.1+cu113 CUDA:0 (Tesla T4, 15110MiB)

Fusing layers... 
YOLOv5s-seg summary: 224 layers, 7611485 parameters, 0 gradients, 26.4 GFLOPs
val: Scanning /content/datasets/coco/val2017... 4952 images, 48 backgrounds, 0 corrupt: 100% 5000/5000 [00:03<00:00, 1361.31it/s]
val: New cache created: /content/datasets/coco/val2017.cache
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100% 157/157 [01:54<00:00,  1.37it/s]
                   all       5000      36335      0.673      0.517      0.566      0.373

# 3. Train

<p align=""><a href="https://roboflow.com/?ref=ultralytics"><img width="1000" src="https://github.com/ultralytics/assets/raw/master/im/integrations-loop.png"/></a></p>
Close the active learning loop by sampling images from your inference conditions with the `roboflow` pip package
<br><br>

Train a YOLOv5s-seg model on the [COCO128](https://www.kaggle.com/ultralytics/coco128) dataset with `--data coco128-seg.yaml`, starting from pretrained `--weights yolov5s-seg.pt`, or from randomly initialized `--weights '' --cfg yolov5s-seg.yaml`.

- **Pretrained [Models](https://github.com/ultralytics/yolov5/tree/master/models)** are downloaded
automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases)
- **[Datasets](https://github.com/ultralytics/yolov5/tree/master/data)** available for autodownload include: [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [COCO128](https://github.com/ultralytics/yolov5/blob/master/data/coco128.yaml), [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [Argoverse](https://github.com/ultralytics/yolov5/blob/master/data/Argoverse.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml), [xView](https://github.com/ultralytics/yolov5/blob/master/data/xView.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [SKU-110K](https://github.com/ultralytics/yolov5/blob/master/data/SKU-110K.yaml).
- **Training Results** are saved to `runs/train-seg/` with incrementing run directories, i.e. `runs/train-seg/exp2`, `runs/train-seg/exp3` etc.
<br><br>

A **Mosaic Dataloader** is used for training which combines 4 images into 1 mosaic.

## Train on Custom Data with Roboflow 🌟 NEW

[Roboflow](https://roboflow.com/?ref=ultralytics) enables you to easily **organize, label, and prepare** a high quality dataset with your own custom data. Roboflow also makes it easy to establish an active learning pipeline, collaborate with your team on dataset improvement, and integrate directly into your model building workflow with the `roboflow` pip package.

- Custom Training Example: [https://blog.roboflow.com/train-yolov5-instance-segmentation-custom-dataset/](https://blog.roboflow.com/train-yolov5-instance-segmentation-custom-dataset/?ref=ultralytics)
- Custom Training Notebook: [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1JTz7kpmHsg-5qwVz2d2IH3AaenI1tv0N?usp=sharing)
<br>

<p align=""><a href="https://roboflow.com/?ref=ultralytics"><img width="480" src="https://robflow-public-assets.s3.amazonaws.com/how-to-train-yolov5-segmentation-annotation.gif"/></a></p>Label images lightning fast (including with model-assisted labeling)

In [ ]:
#@title Select YOLOv5 🚀 logger {run: 'auto'}
logger = 'TensorBoard' #@param ['TensorBoard', 'Comet', 'ClearML']

if logger == 'TensorBoard':
  %load_ext tensorboard
  %tensorboard --logdir runs/train-seg
elif logger == 'Comet':
  %pip install -q comet_ml
  import comet_ml; comet_ml.init()
elif logger == 'ClearML':
  import clearml; clearml.browser_login()

In [14]:
# Train YOLOv5s on COCO128 for 3 epochs
!python train.py --img 640 --batch 16 --epochs 1000 --data coco128.yaml --weights yolov5s.pt --cache ram

train: weights=yolov5s.pt, cfg=, data=coco128.yaml, hyp=data/hyps/hyp.scratch-low.yaml, epochs=1000, batch_size=16, imgsz=640, rect=False, resume=False, nosave=False, noval=False, noautoanchor=False, noplots=False, evolve=None, bucket=, cache=ram, image_weights=False, device=, multi_scale=False, single_cls=False, optimizer=SGD, sync_bn=False, workers=8, project=runs/train, name=exp, exist_ok=False, quad=False, cos_lr=False, label_smoothing=0.0, patience=100, freeze=[0], save_period=-1, seed=0, local_rank=-1, entity=None, upload_dataset=False, bbox_interval=-1, artifact_alias=latest
github: up to date with https://github.com/ultralytics/yolov5 ✅
YOLOv5 🚀 v7.0-33-g1752768 Python-3.8.10 torch-1.13.0+cu117 CUDA:0 (NVIDIA GeForce RTX 3080, 9996MiB)

hyperparameters: lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=0.05, cls=0.5, cls_pw=1.0, obj=1.0, obj_pw=1.0, iou_t=0.2, anchor_t=4.0, fl_gamma=0.0, hsv_h=0.015, hsv_s=0

                 Class     Images  Instances          P          R      mAP50   
                   all         22         41     0.0019       0.31    0.00174   0.000518

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
      7/999      3.62G     0.1009    0.04969     0.0522         16        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41    0.00227      0.353     0.0026    0.00065

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
      8/999      3.62G    0.09738    0.05017    0.05009         21        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41    0.00292       0.45    0.00342   0.000799

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
      9/999      3.62G    0.09323    0.04849    0.05065         17        640: 1
          

     31/999      3.62G    0.07937     0.0499    0.04557         25        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.838      0.111      0.052     0.0135

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
     32/999      3.62G    0.08253    0.04295     0.0459         21        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.241      0.424      0.141     0.0472

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
     33/999      3.62G     0.0716    0.03702    0.04648         14        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.241      0.424      0.141     0.0472

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
     34/99

     56/999      3.62G    0.06586    0.04037     0.0419         19        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41     0.0847      0.324      0.134     0.0573

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
     57/999      3.62G     0.0621    0.04984    0.04291         27        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.312      0.223     0.0963     0.0315

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
     58/999      3.62G    0.06554    0.04085      0.042         21        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.312      0.223     0.0963     0.0315

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
     59/99

     81/999      3.62G    0.05156    0.03915    0.03892         29        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.133      0.475      0.281      0.135

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
     82/999      3.62G    0.05766    0.03823    0.03843         25        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.133      0.475      0.281      0.135

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
     83/999      3.62G    0.05771    0.03538    0.03881         25        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.319       0.36      0.244      0.101

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
     84/99

    106/999      3.62G    0.04527    0.03629     0.0352         27        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.328      0.703       0.52      0.258

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    107/999      3.62G    0.04429    0.02929    0.03328         23        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.563      0.559      0.532      0.281

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    108/999      3.62G    0.04924     0.0315     0.0347         23        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.563      0.559      0.532      0.281

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    109/99

    131/999      3.62G    0.04329    0.02405    0.02861         20        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.316      0.511      0.499      0.203

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    132/999      3.62G    0.05697    0.02824    0.02988         29        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.316      0.511      0.499      0.203

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    133/999      3.62G    0.05584    0.02744    0.03068         19        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.411      0.677      0.604      0.252

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    134/99

    156/999      3.62G     0.0442    0.02314    0.02517         22        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41       0.48      0.845      0.729      0.417

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    157/999      3.62G    0.04267    0.02502    0.02697         21        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.573      0.739      0.731      0.447

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    158/999      3.62G    0.04384    0.03276    0.02606         29        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.573      0.739      0.731      0.447

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    159/99

    181/999      3.62G    0.03256    0.02524    0.02121         19        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.599      0.929      0.822      0.502

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    182/999      3.62G    0.03193    0.02468    0.01907         24        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.599      0.929      0.822      0.502

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    183/999      3.62G    0.03794     0.0249    0.02043         23        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.559      0.951      0.759      0.368

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    184/99

    206/999      3.62G    0.03968    0.01823    0.01789         15        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.537      0.951      0.807      0.462

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    207/999      3.62G    0.04076    0.02307    0.01799         21        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.582      0.931      0.801      0.511

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    208/999      3.62G    0.04342    0.02852    0.01777         30        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.582      0.931      0.801      0.511

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    209/99

    231/999      3.62G    0.03571    0.01744    0.01566         18        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.681      0.964      0.848      0.506

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    232/999      3.62G    0.04421      0.026    0.01688         28        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.681      0.964      0.848      0.506

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    233/999      3.62G    0.03858    0.02129    0.01461         17        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.613      0.919      0.812      0.546

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    234/99

    256/999      3.62G    0.03599     0.0182    0.01316         17        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.635      0.894       0.84      0.508

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    257/999      3.62G    0.03457    0.02572    0.01439         28        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41       0.94      0.731      0.844      0.573

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    258/999      3.62G    0.03288    0.02047    0.01587         24        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41       0.94      0.731      0.844      0.573

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    259/99

    281/999      3.62G    0.03657     0.0233    0.01454         18        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41       0.68      0.971      0.844      0.596

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    282/999      3.62G    0.03101    0.02213    0.01293         25        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41       0.68      0.971      0.844      0.596

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    283/999      3.62G    0.03157    0.02425    0.01308         29        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.612      0.944      0.859      0.588

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    284/99

    306/999      3.62G    0.02668    0.01753     0.0142         20        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.825      0.943       0.93      0.602

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    307/999      3.62G    0.03165    0.01913    0.01457         19        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.823      0.943      0.946      0.639

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    308/999      3.62G    0.02836     0.0201     0.0114         21        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.823      0.943      0.946      0.639

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    309/99

    331/999      3.62G    0.02955    0.01604    0.01241         18        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.831      0.943      0.945      0.674

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    332/999      3.62G    0.02644    0.01665    0.01159         20        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.831      0.943      0.945      0.674

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    333/999      3.62G    0.03148    0.01524    0.01209         15        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.789      0.914      0.929      0.647

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    334/99

    356/999      3.62G    0.02791    0.01837    0.01157         26        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.891      0.971      0.968      0.724

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    357/999      3.62G    0.03021    0.01595    0.01091         14        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.826      0.971      0.948      0.701

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    358/999      3.62G    0.02631    0.01831    0.01329         17        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.826      0.971      0.948      0.701

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    359/99

    381/999      3.62G    0.03558    0.01599    0.01095         15        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41       0.95      0.957      0.968      0.654

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    382/999      3.62G    0.02555    0.02063   0.009148         25        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41       0.95      0.957      0.968      0.654

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    383/999      3.62G    0.02747    0.02218    0.01001         23        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41       0.96      0.965      0.968      0.704

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    384/99

    406/999      3.62G    0.02601    0.01613    0.00801         19        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.958      0.962      0.968      0.741

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    407/999      3.62G    0.02512    0.01706   0.008705         24        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.929      0.943      0.962      0.732

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    408/999      3.62G    0.02388    0.01531   0.008034         15        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.929      0.943      0.962      0.732

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    409/99

    431/999      3.62G     0.0232    0.02043   0.008561         28        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.963      0.971      0.968      0.739

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    432/999      3.62G    0.02599    0.01782   0.007862         16        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.963      0.971      0.968      0.739

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    433/999      3.62G    0.02782    0.01866   0.008139         22        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41       0.97      0.969      0.968      0.653

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    434/99

    456/999      3.62G    0.02472    0.01799   0.005869         21        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.979      0.954      0.968      0.773

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    457/999      3.62G    0.02702    0.01633   0.004673         11        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.981      0.958      0.968      0.703

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    458/999      3.62G    0.02945     0.0163   0.005553         21        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.981      0.958      0.968      0.703

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    459/99

    481/999      3.62G     0.0213    0.01503   0.004385         23        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.946      0.971      0.964      0.759

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    482/999      3.62G    0.02593    0.01654   0.006104         22        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.946      0.971      0.964      0.759

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    483/999      3.62G    0.02383    0.01535   0.005147         18        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.983      0.971      0.968      0.758

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    484/99

    506/999      3.62G     0.0249    0.01805   0.004271         24        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.978      0.971      0.968      0.766

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    507/999      3.62G    0.02417    0.01816   0.004623         21        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.983      0.971      0.968      0.772

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    508/999      3.62G    0.02447    0.01597   0.004979         22        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.983      0.971      0.968      0.772

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    509/99

    531/999      3.62G    0.02761    0.01997   0.005565         23        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.977      0.971      0.968      0.749

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    532/999      3.62G    0.02271    0.01488   0.004288         23        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.977      0.971      0.968      0.749

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    533/999      3.62G     0.0236    0.01636   0.004555         16        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.977      0.971      0.968      0.775

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    534/99

    556/999      3.62G    0.02307    0.01303   0.003188         13        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.991      0.963      0.968      0.781

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    557/999      3.62G    0.02478    0.01546   0.004022         23        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.989      0.966      0.968      0.769

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    558/999      3.62G    0.02332    0.01402   0.003951         14        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.989      0.966      0.968      0.769

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    559/99

    581/999      3.62G    0.02017    0.01687   0.004008         19        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41       0.99      0.971      0.968      0.801

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    582/999      3.62G    0.01966    0.01582   0.003892         22        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41       0.99      0.971      0.968      0.801

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    583/999      3.62G    0.02209    0.01487   0.007599         14        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41       0.99      0.971      0.968       0.78

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    584/99

    606/999      3.62G    0.02238    0.01588    0.00351         19        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41       0.99      0.971      0.968        0.8

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    607/999      3.62G    0.02252    0.01499    0.00348         21        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.989      0.971      0.968      0.773

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    608/999      3.62G    0.02181    0.02058   0.003013         35        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.989      0.971      0.968      0.773

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    609/99

    631/999      3.62G    0.02659    0.01794    0.00383         27        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.988      0.971      0.968      0.791

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    632/999      3.62G    0.02328     0.0171   0.003246         27        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.988      0.971      0.968      0.791

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    633/999      3.62G     0.0207    0.01657    0.00382         26        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.988      0.971      0.968      0.838

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    634/99

    656/999      3.62G    0.02128    0.01553   0.003287         18        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.989      0.971      0.968      0.844

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    657/999      3.62G    0.01776    0.01437   0.003693         22        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.984      0.971      0.968      0.815

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    658/999      3.62G    0.02003    0.01549    0.00246         22        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.984      0.971      0.968      0.815

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    659/99

    681/999      3.62G     0.0208    0.01679   0.003532         27        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.988      0.971      0.968      0.858

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    682/999      3.62G    0.01683    0.01294   0.002597         20        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.988      0.971      0.968      0.858

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    683/999      3.62G    0.01991    0.01429   0.002518         25        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.988      0.971      0.968       0.74

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    684/99

    706/999      3.62G    0.01882    0.01506   0.002296         22        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.989      0.971      0.969      0.808

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    707/999      3.62G    0.01985    0.01189   0.002994         20        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.989      0.971      0.968      0.781

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    708/999      3.62G    0.01929    0.01475   0.002196         23        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.989      0.971      0.968      0.781

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    709/99

    731/999      3.62G    0.01757     0.0145   0.002834         24        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.989      0.971      0.968      0.821

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    732/999      3.62G    0.01639    0.01233    0.00199         20        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.989      0.971      0.968      0.821

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    733/999      3.62G    0.01801    0.01422   0.002942         21        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.988      0.971      0.968      0.821

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    734/99

    756/999      3.62G    0.02001    0.01414   0.003141         18        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41       0.99      0.971      0.969      0.838

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    757/999      3.62G     0.0196    0.02062   0.002784         31        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41       0.99      0.971      0.969      0.862

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    758/999      3.62G    0.01883    0.01286   0.003849         25        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41       0.99      0.971      0.969      0.862

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    759/99

    781/999      3.62G    0.01789    0.01217   0.003596         27        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.988       0.97      0.968      0.837

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    782/999      3.62G    0.01504    0.01326   0.002071         20        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.988       0.97      0.968      0.837

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    783/999      3.62G    0.01836    0.01425   0.005412         25        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.988       0.97      0.968      0.858

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    784/99

    806/999      3.62G    0.01805     0.0164   0.002583         31        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.988      0.971      0.969      0.871

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    807/999      3.62G    0.01903    0.01216   0.003832         22        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.989      0.971      0.969      0.874

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    808/999      3.62G    0.01651    0.01381   0.002683         13        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.989      0.971      0.969      0.874

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    809/99

    831/999      3.62G    0.01657    0.01405   0.002953         31        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.989      0.971      0.969      0.863

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    832/999      3.62G    0.01581    0.01498   0.002499         26        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.989      0.971      0.969      0.863

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    833/999      3.62G    0.01777   0.007933   0.001215          9        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.989      0.971      0.969      0.862

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    834/99

    856/999      3.62G     0.0154    0.01453   0.002812         30        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.989      0.971      0.969       0.88

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    857/999      3.62G    0.01249    0.01369   0.002153         25        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.989      0.971      0.969      0.895

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    858/999      3.62G    0.01247    0.01179   0.001964         24        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.989      0.971      0.969      0.895

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    859/99

    881/999      3.62G    0.01486    0.01315   0.003915         21        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.989      0.971      0.969      0.901

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    882/999      3.62G    0.01393    0.01038   0.002132         19        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.989      0.971      0.969      0.901

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    883/999      3.62G    0.01509    0.01681   0.003158         37        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.989      0.971      0.969      0.899

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    884/99

    906/999      3.62G    0.01423    0.01239   0.001917         23        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.991      0.971      0.969      0.899

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    907/999      3.62G    0.01485    0.01374   0.002265         27        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.991      0.971      0.969      0.916

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    908/999      3.62G    0.01446     0.0132   0.002213         26        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.991      0.971      0.969      0.916

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    909/99

    931/999      3.62G    0.01489    0.01009   0.001621         12        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41       0.99      0.971      0.969      0.929

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    932/999      3.62G    0.01347    0.01508   0.001988         23        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41       0.99      0.971      0.969      0.929

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    933/999      3.62G    0.01226   0.008473    0.00162         14        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41       0.99      0.971      0.969      0.922

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    934/99

    956/999      3.62G    0.01588    0.01391   0.002817         30        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.988      0.971      0.969      0.896

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    957/999      3.62G    0.01303    0.01515   0.002073         29        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.988      0.971      0.969      0.911

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    958/999      3.62G    0.01327   0.009677   0.001425         12        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.988      0.971      0.969      0.911

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    959/99

    981/999      3.62G    0.01344    0.01238   0.001852         23        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.989      0.971      0.969      0.923

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    982/999      3.62G    0.01344    0.00837   0.003299         16        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.989      0.971      0.969      0.923

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    983/999      3.62G    0.01228    0.01108   0.001461         18        640: 1
                 Class     Images  Instances          P          R      mAP50   
                   all         22         41      0.989      0.971      0.969      0.924

      Epoch    GPU_mem   box_loss   obj_loss   cls_loss  Instances       Size
    984/99

# 4. Visualize

## Comet Logging and Visualization 🌟 NEW
[Comet](https://bit.ly/yolov5-readme-comet) is now fully integrated with YOLOv5. Track and visualize model metrics in real time, save your hyperparameters, datasets, and model checkpoints, and visualize your model predictions with [Comet Custom Panels](https://bit.ly/yolov5-colab-comet-panels)! Comet makes sure you never lose track of your work and makes it easy to share results and collaborate across teams of all sizes! 

Getting started is easy:
```shell
pip install comet_ml  # 1. install
export COMET_API_KEY=<Your API Key>  # 2. paste API key
python train.py --img 640 --epochs 3 --data coco128.yaml --weights yolov5s.pt  # 3. train
```

To learn more about all of the supported Comet features for this integration, check out the [Comet Tutorial](https://github.com/ultralytics/yolov5/tree/master/utils/loggers/comet). If you'd like to learn more about Comet, head over to our [documentation](https://bit.ly/yolov5-colab-comet-docs). Get started by trying out the Comet Colab Notebook:
[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1RG0WOQyxlDlo5Km8GogJpIEJlg_5lyYO?usp=sharing)

<img width="1920" alt="yolo-ui" src="https://user-images.githubusercontent.com/26833433/202851203-164e94e1-2238-46dd-91f8-de020e9d6b41.png">

## ClearML Logging and Automation 🌟 NEW

[ClearML](https://cutt.ly/yolov5-notebook-clearml) is completely integrated into YOLOv5 to track your experimentation, manage dataset versions and even remotely execute training runs. To enable ClearML (check cells above):

- `pip install clearml`
- run `clearml-init` to connect to a ClearML server (**deploy your own [open-source server](https://github.com/allegroai/clearml-server)**, or use our [free hosted server](https://cutt.ly/yolov5-notebook-clearml))

You'll get all the great expected features from an experiment manager: live updates, model upload, experiment comparison etc. but ClearML also tracks uncommitted changes and installed packages for example. Thanks to that ClearML Tasks (which is what we call experiments) are also reproducible on different machines! With only 1 extra line, we can schedule a YOLOv5 training task on a queue to be executed by any number of ClearML Agents (workers).

You can use ClearML Data to version your dataset and then pass it to YOLOv5 simply using its unique ID. This will help you keep track of your data without adding extra hassle. Explore the [ClearML Tutorial](https://github.com/ultralytics/yolov5/tree/master/utils/loggers/clearml) for details!

<a href="https://cutt.ly/yolov5-notebook-clearml">
<img alt="ClearML Experiment Management UI" src="https://github.com/thepycoder/clearml_screenshots/raw/main/scalars.jpg" width="1280"/></a>

## Local Logging

Training results are automatically logged with [Tensorboard](https://www.tensorflow.org/tensorboard) and [CSV](https://github.com/ultralytics/yolov5/pull/4148) loggers to `runs/train`, with a new experiment directory created for each new training as `runs/train/exp2`, `runs/train/exp3`, etc.

This directory contains train and val statistics, mosaics, labels, predictions and augmentated mosaics, as well as metrics and charts including precision-recall (PR) curves and confusion matrices. 

<img alt="Local logging results" src="https://user-images.githubusercontent.com/26833433/183222430-e1abd1b7-782c-4cde-b04d-ad52926bf818.jpg" width="1280"/>


# Environments

YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://bit.ly/yolov5-paperspace-notebook"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/GCP-Quickstart)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/AWS-Quickstart)
- **Docker Image**. See [Docker Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/Docker-Quickstart) <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>


# Status

![YOLOv5 CI](https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml/badge.svg)

If this badge is green, all [YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions) Continuous Integration (CI) tests are currently passing. CI tests verify correct operation of YOLOv5 training ([train.py](https://github.com/ultralytics/yolov5/blob/master/train.py)), testing ([val.py](https://github.com/ultralytics/yolov5/blob/master/val.py)), inference ([detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py)) and export ([export.py](https://github.com/ultralytics/yolov5/blob/master/export.py)) on macOS, Windows, and Ubuntu every 24 hours and on every commit.


# Appendix

Additional content below.

In [ ]:
# YOLOv5 PyTorch HUB Inference (DetectionModels only)
import torch

model = torch.hub.load('ultralytics/yolov5', 'yolov5s-seg')  # yolov5n - yolov5x6 or custom
im = 'https://ultralytics.com/images/zidane.jpg'  # file, Path, PIL.Image, OpenCV, nparray, list
results = model(im)  # inference
results.print()  # or .show(), .save(), .crop(), .pandas(), etc.